File: src/transport.ts
import type { LogRecord } from "./types.ts";

const activeTransports = new Set<Transport>();
let shutdownRegistered = false;

const SHUTDOWN_DRAIN_TIMEOUT_MS = 5000;

function registerShutdownHandlers(): void {
	if (shutdownRegistered) return;
	shutdownRegistered = true;

	if (typeof process === "undefined") return;

	// Drain everything: pending buffered batches AND any sendBatch already
	// in flight. Bound the wait so a hung HTTP request doesn't pin the
	// process forever — at SHUTDOWN_DRAIN_TIMEOUT_MS we give up and let
	// the process exit, dropping any still-in-flight batches.
	const drainAll = async () => {
		const flushes = [...activeTransports].map((t) => t.flush());
		const inflight = [...activeTransports].flatMap((t) => t.getInflight());
		const pending = Promise.allSettled([...flushes, ...inflight]);
		const timeout = new Promise<void>((resolve) => {
			const t = setTimeout(resolve, SHUTDOWN_DRAIN_TIMEOUT_MS);
			if (typeof t === "object" && "unref" in t) t.unref();
		});
		await Promise.race([pending, timeout]);
	};

	const shutdownFlush = () => {
		// Pin the event loop while drain is running; clear it so the
		// process can exit immediately afterwards. Bun/Node would otherwise
		// have nothing keeping it alive once timers/intervals are unref'd.
		const keepAlive = setInterval(() => {}, 1000);
		drainAll().finally(() => clearInterval(keepAlive));
	};
	process.on("SIGINT", shutdownFlush);
	process.on("SIGTERM", shutdownFlush);
	// beforeExit fires when the loop is about to drain naturally — last
	// chance to flush leftover buffered logs that batched-but-never-sent.
	// Guard against re-entry: drainAll's async work re-arms the loop,
	// which would re-fire beforeExit and recurse forever.
	let beforeExitFired = false;
	process.on("beforeExit", () => {
		if (beforeExitFired) return;
		beforeExitFired = true;
		void drainAll();
	});
}

export class Transport {
	private buffer: LogRecord[] = [];
	private timer: ReturnType<typeof setInterval> | null = null;
	private url: string;
	private auth: string | undefined;
	private batchSize: number;
	private flushInterval: number;
	private maxBufferSize: number;
	private flushing = false;
	private pendingFlush = false;
	private destroyed = false;
	private onError: ((error: Error, batch: LogRecord[]) => void) | undefined;
	private inflight = new Set<Promise<void>>();

	constructor(options: {
		url: string;
		auth?: string;
		batchSize?: number;
		flushInterval?: number;
		maxBufferSize?: number;
		onError?: (error: Error, batch: LogRecord[]) => void;
	}) {
		this.url = options.url;
		this.auth = options.auth;
		this.batchSize = options.batchSize ?? 50;
		this.flushInterval = options.flushInterval ?? 5000;
		this.maxBufferSize = options.maxBufferSize ?? 10000;
		this.onError = options.onError;

		this.timer = setInterval(() => {
			void this.flush();
		}, this.flushInterval);
		this.timer.unref();

		activeTransports.add(this);
		registerShutdownHandlers();
	}

	send(record: LogRecord): void {
		if (this.destroyed) return;
		if (this.buffer.length >= this.maxBufferSize) {
			this.buffer.splice(0, Math.floor(this.maxBufferSize * 0.1));
		}
		this.buffer.push(record);
		if (this.buffer.length >= this.batchSize) {
			void this.flush();
		}
	}

	async flush(): Promise<void> {
		if (this.flushing) {
			this.pendingFlush = true;
			return;
		}
		if (this.buffer.length === 0) return;

		this.flushing = true;
		this.pendingFlush = false;

		const batch = this.buffer.splice(0);

		const sendPromise = this.sendBatch(batch);
		this.inflight.add(sendPromise);
		try {
			await sendPromise;
		} finally {
			this.inflight.delete(sendPromise);
			this.flushing = false;
			if (!this.destroyed && this.pendingFlush && this.buffer.length > 0) {
				this.pendingFlush = false;
				void this.flush();
			}
		}
	}

	/** Snapshot of in-flight sendBatch promises for the shutdown path. */
	getInflight(): Promise<void>[] {
		return [...this.inflight];
	}

	private async sendBatch(batch: LogRecord[]): Promise<void> {
		if (typeof fetch === "undefined") {
			console.warn(`[relog.sh] fetch is not available, dropping ${batch.length} log(s)`);
			return;
		}

		const headers: Record<string, string> = {
			"Content-Type": "application/json",
		};
		if (this.auth) {
			headers["Authorization"] = `Bearer ${this.auth}`;
		}

		let lastError: Error | undefined;

		for (let attempt = 0; attempt < 3; attempt++) {
			if (this.destroyed) return;

			try {
				const response = await fetch(`${this.url}/ingest`, {
					method: "POST",
					headers,
					body: JSON.stringify(batch),
				});

				if (response.ok) return;

				if (response.status === 429) {
					const retryAfter = response.headers.get("Retry-After");
					const waitMs = retryAfter ? Number.parseInt(retryAfter, 10) * 1000 : 10_000;
					lastError = new Error("Rate limited (429)");
					if (attempt < 2 && !this.destroyed) {
						await new Promise<void>((r) => {
							const t = setTimeout(r, waitMs);
							if (typeof t === "object" && "unref" in t) t.unref();
						});
					}
					continue;
				}

				if (response.status < 500) {
					lastError = new Error(`Ingest rejected: ${response.status}`);
					break;
				}
				lastError = new Error(`Ingest failed: ${response.status}`);
			} catch (err) {
				lastError = err instanceof Error ? err : new Error("Network error");
			}

			if (attempt < 2 && !this.destroyed) {
				const delay = Math.min(2 ** attempt * 500, 5000);
				await new Promise<void>((r) => {
					const t = setTimeout(r, delay);
					if (typeof t === "object" && "unref" in t) t.unref();
				});
			}
		}

		if (lastError) {
			if (this.onError) {
				try {
					this.onError(lastError, batch);
				} catch {
					// onError should not throw
				}
			} else {
				console.warn(`[relog.sh] Failed to send ${batch.length} log(s): ${lastError.message}`);
			}
		}
	}

	destroy(): void {
		this.destroyed = true;
		if (this.timer) {
			clearInterval(this.timer);
			this.timer = null;
		}
		activeTransports.delete(this);
	}
}

File: python/src/relog/_transport.py
from __future__ import annotations

import atexit
import math
import sys
import threading
import time
from typing import TYPE_CHECKING, Any

import httpx

if TYPE_CHECKING:
    from ._types import LogRecord

_active_transports: set[Transport] = set()
_atexit_registered = False


def _register_atexit() -> None:
    global _atexit_registered
    if _atexit_registered:
        return
    _atexit_registered = True

    def _flush_all() -> None:
        for t in list(_active_transports):
            try:
                t.flush()
            except Exception:
                pass

    atexit.register(_flush_all)


class Transport:
    def __init__(
        self,
        *,
        url: str,
        auth: str | None = None,
        batch_size: int = 50,
        flush_interval: float = 5.0,
        max_buffer_size: int = 10_000,
        on_error: Any | None = None,
    ) -> None:
        self._url = url
        self._auth = auth
        self._batch_size = batch_size
        self._flush_interval = flush_interval
        self._max_buffer_size = max_buffer_size
        self._on_error = on_error

        self._buffer: list[LogRecord] = []
        self._lock = threading.Lock()
        self._flushing = False
        self._pending_flush = False
        self._destroyed = False
        self._client: httpx.Client | None = None

        self._stop_event = threading.Event()
        self._timer_thread = threading.Thread(target=self._timer_loop, daemon=True)
        self._timer_thread.start()

        _active_transports.add(self)
        _register_atexit()

    def _get_client(self) -> httpx.Client:
        if self._client is None:
            self._client = httpx.Client(timeout=30.0)
        return self._client

    def _timer_loop(self) -> None:
        while not self._stop_event.wait(self._flush_interval):
            try:
                self.flush()
            except Exception:
                pass

    def send(self, record: LogRecord) -> None:
        if self._destroyed:
            return
        with self._lock:
            if len(self._buffer) >= self._max_buffer_size:
                drop = math.floor(self._max_buffer_size * 0.1)
                del self._buffer[:drop]
            self._buffer.append(record)
            should_flush = len(self._buffer) >= self._batch_size
        if should_flush:
            self.flush()

    def flush(self) -> None:
        with self._lock:
            if self._flushing:
                self._pending_flush = True
                return
            if not self._buffer:
                return
            self._flushing = True
            self._pending_flush = False
            batch = list(self._buffer)
            self._buffer.clear()

        try:
            self._send_batch(batch)
        finally:
            with self._lock:
                self._flushing = False
                needs_reflush = (
                    not self._destroyed
                    and self._pending_flush
                    and len(self._buffer) > 0
                )
            if needs_reflush:
                self.flush()

    def _send_batch(self, batch: list[LogRecord]) -> None:
        headers: dict[str, str] = {"Content-Type": "application/json"}
        if self._auth:
            headers["Authorization"] = f"Bearer {self._auth}"

        payload = [r.to_dict() for r in batch]
        last_error: Exception | None = None
        client = self._get_client()

        for attempt in range(3):
            if self._destroyed:
                return

            try:
                response = client.post(
                    f"{self._url}/ingest",
                    json=payload,
                    headers=headers,
                )

                if response.is_success:
                    return

                if response.status_code == 429:
                    retry_after = response.headers.get("Retry-After")
                    wait = int(retry_after) if retry_after else 10
                    last_error = Exception("Rate limited (429)")
                    if attempt < 2 and not self._destroyed:
                        time.sleep(wait)
                    continue

                if response.status_code < 500:
                    last_error = Exception(
                        f"Ingest rejected: {response.status_code}"
                    )
                    break

                last_error = Exception(f"Ingest failed: {response.status_code}")
            except Exception as exc:
                last_error = exc

            if attempt < 2 and not self._destroyed:
                delay = min(2**attempt * 0.5, 5.0)
                time.sleep(delay)

        if last_error:
            if self._on_error:
                try:
                    self._on_error(last_error, batch)
                except Exception:
                    pass
            else:
                print(
                    f"[relog] Failed to send {len(batch)} log(s): {last_error}",
                    file=sys.stderr,
                )

    def destroy(self) -> None:
        self._destroyed = True
        self._stop_event.set()
        _active_transports.discard(self)
        if self._client:
            try:
                self._client.close()
            except Exception:
                pass
            self._client = None
